File: Bulwark.Core/Engine/BaselineAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Bulwark.Core.Models;

namespace Bulwark.Core.Engine;

/// <summary>
/// 行为基线 / 异常检测器(独创·有状态画像)。
///
/// 现有检测多为「规则 / 特征 / 时序」驱动:它们描述"什么样的行为是坏的"。
/// 本分析器换一个角度 —— 为每个程序建立**正常行为画像(基线)**,描述"这个程序平时怎么做",
/// 当它出现**显著偏离自身历史**的行为时升分。三个维度:
///   · 子进程:某程序平时派生哪些子进程(按子进程映像名);
///   · 外联目标:某程序平时连向哪些远端主机;
///   · 写入目录:某程序平时往哪些目录写文件。
///
/// 典型价值:一个一直只连自家更新服务器的程序,某天突然连向陌生主机并派生 powershell,
/// 即便单看每一步都不命中硬规则,"偏离自身基线"也是有意义的可疑信号(被劫持 / 被注入 /
/// 供应链投毒 / DLL 侧载后行为突变)。
///
/// 严守全局低误报原则:
///   1) **学习期**:某维度观测次数未达 <see cref="MinObservationsToEstablish"/> 前只学习、不评分;
///   2) 偏离基线产出的恒为**软信号**(SoftSignal),<b>单独绝不</b>置位
///      <see cref="SecurityEvent.HasThreatIndicator"/> 或直接处置,仅累加 RiskScore,
///      需与硬指标互证后由调用方升格;
///   3) **高基数豁免**:浏览器 / 下载器这类天然连接/派生海量不同目标的程序,基数超过
///      <see cref="PromiscuousThreshold"/> 后停止对该维度评分,避免持续误报。
///
/// 状态可通过 <see cref="Export"/> / <see cref="Import"/> 快照持久化,跨重启保留画像。
/// 线程安全:单锁串行化,带容量 / 过期治理。
/// </summary>
public sealed class BaselineAnalyzer
{
    private readonly object _gate = new();
    private readonly Dictionary<string, Profile> _profiles = new(StringComparer.OrdinalIgnoreCase);

    private readonly int _minObs;
    private readonly int _promiscuous;
    private readonly int _maxProfiles;
    private readonly int _maxSetSize;

    /// <summary>某维度达到此观测次数后才视为「基线已建立」,之前只学习不评分。</summary>
    public int MinObservationsToEstablish => _minObs;

    /// <summary>某维度的已知集合大小超过此值视为「高基数(promiscuous)」,停止对该维度评分。</summary>
    public int PromiscuousThreshold => _promiscuous;

    public BaselineAnalyzer(
        int minObservationsToEstablish = 12,
        int promiscuousThreshold = 60,
        int maxProfiles = 8192,
        int maxSetSize = 256)
    {
        _minObs = Math.Max(4, minObservationsToEstablish);
        _promiscuous = Math.Max(16, promiscuousThreshold);
        _maxProfiles = Math.Max(128, maxProfiles);
        _maxSetSize = Math.Max(32, maxSetSize);
    }

    /// <summary>
    /// 观测一个事件,更新对应程序画像并研判是否偏离基线。
    /// 返回 (Score, Reasons, IsDeviation)。Score 永远是软信号贡献,调用方决定证据类别 / 是否升格。
    /// 对无法归因到稳定程序身份的事件(空路径等)返回零分。
    /// </summary>
    public (int Score, List<string> Reasons, bool IsDeviation) Observe(SecurityEvent e)
    {
        var reasons = new List<string>();
        if (e is null) return (0, reasons, false);

        var now = e.TimestampUtc == default ? DateTime.UtcNow : e.TimestampUtc;

        switch (e.Type)
        {
            case EventType.ProcessCreate:
            {
                // 子进程派生:画像归属于「父进程」,记录的值是「子进程映像名」。
                string parent = NormalizeProgram(e.ParentPath);
                string child = SafeName(e.ActorPath);
                if (parent.Length == 0 || child.Length == 0) return (0, reasons, false);
                return Score(parent, Dim.Child, child, now,
                    deviationScore: 20,
                    deviationText: $"程序 {ShortName(parent)} 首次派生子进程 {child}(偏离历史行为基线)");
            }
            case EventType.NetworkConnect:
            {
                string actor = NormalizeProgram(e.ActorPath);
                string host = NormalizeRemote(e.Target);
                if (actor.Length == 0 || host.Length == 0 || host == "?") return (0, reasons, false);
                return Score(actor, Dim.Host, host, now,
                    deviationScore: 15,
                    deviationText: $"程序 {ShortName(actor)} 首次外联到 {host}(偏离历史外联基线)");
            }
            case EventType.FileWrite:
            case EventType.FileDelete:
            {
                string actor = NormalizeProgram(e.ActorPath);
                string dir = DirOf(e.Target);
                if (actor.Length == 0 || dir.Length == 0) return (0, reasons, false);
                return Score(actor, Dim.WriteDir, dir, now,
                    deviationScore: 12,
                    deviationText: $"程序 {ShortName(actor)} 首次写入目录 {dir}(偏离历史写入基线)");
            }
            default:
                return (0, reasons, false);
        }
    }

    private (int Score, List<string> Reasons, bool IsDeviation) Score(
        string programKey, Dim dim, string value, DateTime now, int deviationScore, string deviationText)
    {
        var reasons = new List<string>();
        lock (_gate)
        {
            if (!_profiles.TryGetValue(programKey, out var p))
            {
                p = new Profile { FirstSeenUtc = now };
                _profiles[programKey] = p;
            }
            p.LastSeenUtc = now;

            var set = p.SetFor(dim);
            int obs = p.ObsFor(dim);
            bool established = obs >= _minObs;
            bool promiscuous = set.Count >= _promiscuous;
            bool known = set.Contains(value);

            // 学习:计数 +1,加入已知集合(带容量上限,避免无界膨胀)。
            p.IncObs(dim);
            if (!known && set.Count < _maxSetSize) set.Add(value);

            EvictIfNeeded(now);

            // 评分前置条件:基线已建立、非高基数程序、且本次是「新值」。
            if (!established || promiscuous || known)
                return (0, reasons, false);

            reasons.Add(deviationText);
            return (Math.Min(deviationScore, 100), reasons, true);
        }
    }

    /// <summary>进程退出时无需清理:画像按程序身份(路径)而非 PID 聚合,跨进程实例累积。</summary>
    public int TrackedProgramCount
    {
        get { lock (_gate) return _profiles.Count; }
    }

    private void EvictIfNeeded(DateTime now)
    {
        if (_profiles.Count <= _maxProfiles) return;
        var oldest = _profiles.OrderBy(kv => kv.Value.LastSeenUtc)
            .Take(_profiles.Count - _maxProfiles)
            .Select(kv => kv.Key).ToList();
        foreach (var k in oldest) _profiles.Remove(k);
    }

    // ===== 持久化快照 =====

    /// <summary>导出当前所有画像为可序列化快照(供宿主落盘)。</summary>
    public BaselineSnapshot Export()
    {
        lock (_gate)
        {
            return new BaselineSnapshot
            {
                Programs = _profiles.Select(kv => new BaselineProgram
                {
                    Key = kv.Key,
                    FirstSeenUtc = kv.Value.FirstSeenUtc,
                    LastSeenUtc = kv.Value.LastSeenUtc,
                    ChildObs = kv.Value.ChildObs,
                    HostObs = kv.Value.HostObs,
                    WriteObs = kv.Value.WriteObs,
                    Children = kv.Value.Children.ToList(),
                    Hosts = kv.Value.Hosts.ToList(),
                    WriteDirs = kv.Value.WriteDirs.ToList(),
                }).ToList()
            };
        }
    }

    /// <summary>从快照恢复画像(覆盖当前状态)。容错:空快照 / 空字段安全跳过。</summary>
    public void Import(BaselineSnapshot? snapshot)
    {
        if (snapshot?.Programs is null) return;
        lock (_gate)
        {
            _profiles.Clear();
            foreach (var prog in snapshot.Programs)
            {
                if (string.IsNullOrEmpty(prog.Key)) continue;
                var p = new Profile
                {
                    FirstSeenUtc = prog.FirstSeenUtc == default ? DateTime.UtcNow : prog.FirstSeenUtc,
                    LastSeenUtc = prog.LastSeenUtc == default ? DateTime.UtcNow : prog.LastSeenUtc,
                    ChildObs = Math.Max(0, prog.ChildObs),
                    HostObs = Math.Max(0, prog.HostObs),
                    WriteObs = Math.Max(0, prog.WriteObs),
                };
                if (prog.Children is not null) foreach (var c in prog.Children) p.Children.Add(c);
                if (prog.Hosts is not null) foreach (var h in prog.Hosts) p.Hosts.Add(h);
                if (prog.WriteDirs is not null) foreach (var d in prog.WriteDirs) p.WriteDirs.Add(d);
                _profiles[prog.Key] = p;
            }
        }
    }

    // ===== 归一化辅助 =====

    private static string NormalizeProgram(string? path)
    {
        if (string.IsNullOrWhiteSpace(path)) return string.Empty;
        return path.Trim().ToLowerInvariant().Replace('/', '\\');
    }

    private static string ShortName(string programKey)
    {
        try { return System.IO.Path.GetFileName(programKey); }
        catch { return programKey; }
    }

    private static string SafeName(string? path)
    {
        if (string.IsNullOrEmpty(path)) return string.Empty;
        try { return System.IO.Path.GetFileName(path).ToLowerInvariant(); }
        catch { return path.ToLowerInvariant(); }
    }

    private static string DirOf(string? target)
    {
        if (string.IsNullOrWhiteSpace(target)) return string.Empty;
        try
        {
            var dir = System.IO.Path.GetDirectoryName(target.Trim());
            if (string.IsNullOrEmpty(dir)) return string.Empty;
            return dir.ToLowerInvariant().Replace('/', '\\');
        }
        catch { return string.Empty; }
    }

    private static string NormalizeRemote(string? target)
    {
        if (string.IsNullOrEmpty(target)) return "?";
        var t = target.Trim().ToLowerInvariant();
        int colon = t.LastIndexOf(':');
        if (colon > 0 && colon < t.Length - 1 && t.Skip(colon + 1).All(char.IsDigit))
            t = t.Substring(0, colon);
        return t;
    }

    private enum Dim { Child, Host, WriteDir }

    /// <summary>单个程序的行为画像。</summary>
    private sealed class Profile
    {
        public DateTime FirstSeenUtc;
        public DateTime LastSeenUtc;
        public int ChildObs;
        public int HostObs;
        public int WriteObs;
        public readonly HashSet<string> Children = new(StringComparer.OrdinalIgnoreCase);
        public readonly HashSet<string> Hosts = new(StringComparer.OrdinalIgnoreCase);
        public readonly HashSet<string> WriteDirs = new(StringComparer.OrdinalIgnoreCase);

        public HashSet<string> SetFor(Dim d) => d switch
        {
            Dim.Child => Children,
            Dim.Host => Hosts,
            _ => WriteDirs
        };

        public int ObsFor(Dim d) => d switch
        {
            Dim.Child => ChildObs,
            Dim.Host => HostObs,
            _ => WriteObs
        };

        public void IncObs(Dim d)
        {
            switch (d)
            {
                case Dim.Child: ChildObs++; break;
                case Dim.Host: HostObs++; break;
                default: WriteObs++; break;
            }
        }
    }
}

/// <summary>行为基线持久化快照(整体)。</summary>
public sealed class BaselineSnapshot
{
    public List<BaselineProgram> Programs { get; set; } = new();
}

/// <summary>单个程序画像的可序列化形态。</summary>
public sealed class BaselineProgram
{
    public string Key { get; set; } = string.Empty;
    public DateTime FirstSeenUtc { get; set; }
    public DateTime LastSeenUtc { get; set; }
    public int ChildObs { get; set; }
    public int HostObs { get; set; }
    public int WriteObs { get; set; }
    public List<string> Children { get; set; } = new();
    public List<string> Hosts { get; set; } = new();
    public List<string> WriteDirs { get; set; } = new();
}

File: cpp/shared/src/engine/BaselineAnalyzer.cpp
#include "bulwark/engine/BaselineAnalyzer.h"
#include <QRegularExpression>
#include <QStringList>
#include <QVector>
#include <algorithm>

namespace bulwark::engine {
using detail::u;
using detail::fileNameLower;

namespace {

//
// 把路径里【每次都不一样】的目录名折叠成固定占位符。
//
// 为什么必需:基线按完整路径字符串做 key,而大量正常程序每次操作都新建一个随机名目录,
// 于是「首次写入该目录」永远成立 —— 基线【永远收敛不了】,每一次正常运行都产出一条
// 「偏离历史基线」。实测 360 浏览器自己的更新器 360ceupdate.exe 就是这样,它每次更新都写
//     ...\Chrome\User Data\v3update\download\{17dbc974-e3ab-475b-9f84-e8a831c494a2}
// GUID 每次不同,累计刷出 13 条偏离证据,把风险分一路顶到 70 并被拦。
//
// 折叠之后这 13 条塌成同一条 ...\v3update\download\{guid},第二次更新起就是「已知目录」。
// 这不但消除误报,还让判据更有意义:真正值得关注的是「写进了哪一类位置」,而不是那串随机名。
//
// 覆盖实测见过的几种易变命名(都换成语义占位符,彼此不混淆):
//   {8-4-4-4-12} / 裸 GUID        -> {guid}
//   _MEI<数字>(PyInstaller 解包)  -> {pyinstaller}
//   is-XXXXX.tmp(Inno Setup)      -> {innosetup}
//   ns<随机>.tmp(NSIS)            -> {nsis}
//   纯十六进制且长度 >= 8           -> {hex}
//   纯数字且长度 >= 4(按 PID 命名)  -> {num}
//
QString collapseVolatileSegments(const QString& lowerPath) {
    static const QRegularExpression reGuidBraced(
        QStringLiteral("^\\{?[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\\}?$"));
    static const QRegularExpression reMei(QStringLiteral("^_mei\\d+$"));
    static const QRegularExpression reInno(QStringLiteral("^is-[0-9a-z]{5,}\\.tmp$"));
    static const QRegularExpression reNsis(QStringLiteral("^ns[0-9a-z]{3,}\\.tmp$"));
    static const QRegularExpression reHex(QStringLiteral("^[0-9a-f]{8,}$"));
    static const QRegularExpression reNum(QStringLiteral("^\\d{4,}$"));

    QStringList segs = lowerPath.split(QLatin1Char('\\'));
    for (QString& s : segs) {
        if (s.isEmpty()) continue;
        if (reGuidBraced.match(s).hasMatch())   { s = QStringLiteral("{guid}");        continue; }
        if (reMei.match(s).hasMatch())          { s = QStringLiteral("{pyinstaller}"); continue; }
        if (reInno.match(s).hasMatch())         { s = QStringLiteral("{innosetup}");   continue; }
        if (reNsis.match(s).hasMatch())         { s = QStringLiteral("{nsis}");        continue; }
        if (reHex.match(s).hasMatch())          { s = QStringLiteral("{hex}");         continue; }
        if (reNum.match(s).hasMatch())          { s = QStringLiteral("{num}");         continue; }
    }
    return segs.join(QLatin1Char('\\'));
}

QString normalizeProgram(const QString& path) {
    if (path.trimmed().isEmpty()) return QString();
    QString p = path.trimmed().toLower();
    p.replace(QLatin1Char('/'), QLatin1Char('\\'));
    // 主体路径也要折叠:PyInstaller 打包的程序每次运行都在 _MEI<pid> 下解包,不折叠的话
    // 每启动一次就新建一份 profile,基线同样永远建立不起来(而且白占 maxProfiles_ 额度)。
    return collapseVolatileSegments(p);
}
QString dirOf(const QString& target) {
    if (target.trimmed().isEmpty()) return QString();
    QString t = target.trimmed();
    t.replace(QLatin1Char('/'), QLatin1Char('\\'));
    const int slash = t.lastIndexOf(QLatin1Char('\\'));
    if (slash < 0) return QString();
    return collapseVolatileSegments(t.left(slash).toLower());
}
QString normalizeRemote(const QString& target) {
    if (target.isEmpty()) return QStringLiteral("?");
    QString t = target.trimmed().toLower();
    const int colon = t.lastIndexOf(QLatin1Char(':'));
    if (colon > 0 && colon < t.size() - 1) {
        bool allDigits = true;
        for (int i = colon + 1; i < t.size(); ++i)
            if (!t.at(i).isDigit()) { allDigits = false; break; }
        if (allDigits) t = t.left(colon);
    }
    return t;
}
QString shortName(const QString& programKey) { return fileNameLower(programKey); }
} // namespace

QSet<QString>& BaselineAnalyzer::Profile::setFor(Dim d) {
    switch (d) {
        case Dim::Child: return children;
        case Dim::Host:  return hosts;
        default:         return writeDirs;
    }
}
int BaselineAnalyzer::Profile::obsFor(Dim d) const {
    switch (d) {
        case Dim::Child: return childObs;
        case Dim::Host:  return hostObs;
        default:         return writeObs;
    }
}
void BaselineAnalyzer::Profile::incObs(Dim d) {
    switch (d) {
        case Dim::Child: ++childObs; break;
        case Dim::Host:  ++hostObs; break;
        default:         ++writeObs; break;
    }
}

BaselineAnalyzer::BaselineAnalyzer(int minObservationsToEstablish, int promiscuousThreshold,
                                   int maxProfiles, int maxSetSize) {
    minObs_ = qMax(4, minObservationsToEstablish);
    promiscuous_ = qMax(16, promiscuousThreshold);
    maxProfiles_ = qMax(128, maxProfiles);
    maxSetSize_ = qMax(32, maxSetSize);
}

BaselineAnalyzer::Result BaselineAnalyzer::observe(const SecurityEvent& e) {
    const QDateTime now = e.timestampUtc.isValid() ? e.timestampUtc : nowUtc();

    switch (e.type) {
        case EventType::ProcessCreate: {
            const QString parent = normalizeProgram(e.parentPath);
            const QString child = fileNameLower(e.actorPath);
            if (parent.isEmpty() || child.isEmpty()) return Result{};
            return score(parent, Dim::Child, child, now, 20,
                         u("程序 ") + shortName(parent) + u(" 首次派生子进程 ") + child +
                         u("(偏离历史行为基线)"));
        }
        case EventType::NetworkConnect: {
            const QString actor = normalizeProgram(e.actorPath);
            const QString host = normalizeRemote(e.target);
            if (actor.isEmpty() || host.isEmpty() || host == QLatin1String("?")) return Result{};
            return score(actor, Dim::Host, host, now, 15,
                         u("程序 ") + shortName(actor) + u(" 首次外联到 ") + host +
                         u("(偏离历史外联基线)"));
        }
        case EventType::FileWrite:
        case EventType::FileDelete: {
            const QString actor = normalizeProgram(e.actorPath);
            const QString dir = dirOf(e.target);
            if (actor.isEmpty() || dir.isEmpty()) return Result{};
            return score(actor, Dim::WriteDir, dir, now, 12,
                         u("程序 ") + shortName(actor) + u(" 首次写入目录 ") + dir +
                         u("(偏离历史写入基线)"));
        }
        default:
            return Result{};
    }
}

BaselineAnalyzer::Result BaselineAnalyzer::score(const QString& programKey, Dim dim, const QString& value,
                                                 const QDateTime& now, int deviationScore,
                                                 const QString& deviationText) {
    QMutexLocker locker(&gate_);

    auto it = profiles_.find(programKey);
    if (it == profiles_.end()) {
        Profile p;
        p.firstSeenUtc = now;
        it = profiles_.insert(programKey, p);
    }
    Profile& p = it.value();
    p.lastSeenUtc = now;

    QSet<QString>& set = p.setFor(dim);
    const int obs = p.obsFor(dim);
    const bool established = obs >= minObs_;
    const bool promiscuous = set.size() >= promiscuous_;
    const bool known = set.contains(value);

    p.incObs(dim);
    if (!known && set.size() < maxSetSize_) set.insert(value);

    evictIfNeeded(now);

    if (!established || promiscuous || known) return Result{};

    Result r;
    r.reasons << deviationText;
    r.score = qMin(deviationScore, 100);
    r.deviation = true;
    return r;
}

int BaselineAnalyzer::trackedProgramCount() {
    QMutexLocker locker(&gate_);
    return profiles_.size();
}

void BaselineAnalyzer::evictIfNeeded(const QDateTime& now) {
    Q_UNUSED(now);
    if (profiles_.size() <= maxProfiles_) return;
    QVector<QPair<QDateTime, QString>> byAge;
    byAge.reserve(profiles_.size());
    for (auto it = profiles_.constBegin(); it != profiles_.constEnd(); ++it)
        byAge.append({ it.value().lastSeenUtc, it.key() });
    std::sort(byAge.begin(), byAge.end(),
              [](const auto& a, const auto& b) { return a.first < b.first; });
    const int toRemove = profiles_.size() - maxProfiles_;
    for (int i = 0; i < toRemove; ++i) profiles_.remove(byAge.at(i).second);
}

BaselineSnapshot BaselineAnalyzer::exportSnapshot() {
    QMutexLocker locker(&gate_);
    BaselineSnapshot snap;
    snap.programs.reserve(profiles_.size());
    for (auto it = profiles_.constBegin(); it != profiles_.constEnd(); ++it) {
        const Profile& p = it.value();
        BaselineProgram bp;
        bp.key = it.key();
        bp.firstSeenUtc = p.firstSeenUtc;
        bp.lastSeenUtc = p.lastSeenUtc;
        bp.childObs = p.childObs;
        bp.hostObs = p.hostObs;
        bp.writeObs = p.writeObs;
        bp.children = QStringList(p.children.values());
        bp.hosts = QStringList(p.hosts.values());
        bp.writeDirs = QStringList(p.writeDirs.values());
        snap.programs.append(bp);
    }
    return snap;
}

void BaselineAnalyzer::importSnapshot(const BaselineSnapshot& snapshot) {
    QMutexLocker locker(&gate_);
    profiles_.clear();
    // 局部变量改名以避让 bulwark::nowUtc():原名恰好与函数同名,会构成自引用初始化。
    const QDateTime importedAt = bulwark::nowUtc();
    for (const BaselineProgram& prog : snapshot.programs) {
        if (prog.key.isEmpty()) continue;
        Profile p;
        p.firstSeenUtc = prog.firstSeenUtc.isValid() ? prog.firstSeenUtc : importedAt;
        p.lastSeenUtc = prog.lastSeenUtc.isValid() ? prog.lastSeenUtc : importedAt;
        p.childObs = qMax(0, prog.childObs);
        p.hostObs = qMax(0, prog.hostObs);
        p.writeObs = qMax(0, prog.writeObs);
        for (const QString& c : prog.children) p.children.insert(c);
        for (const QString& h : prog.hosts) p.hosts.insert(h);
        for (const QString& d : prog.writeDirs) p.writeDirs.insert(d);
        profiles_.insert(prog.key, p);
    }
}

} // namespace bulwark::engine
